team_check.py

```diff
@@ -0,0 +1,78 @@
+import pandas as pd
+from github3 import login
+import configparser
+import numpy as np
+
+
+### Get team members
+config = configparser.ConfigParser()
+config.read('config')
+config = config['DEFAULT']
+
+username = config['username']
+token = config['token']
+organization = config['organization']
+pairs = config['pairs']
+roster = config['roster']
+
+with open(token, 'r') as f:
+    token = f.readline().strip()
+    
+github = login(username,token=token)
+memberships = github.organization_memberships()
+course = None
+
+for membership in memberships:
+    if membership.organization.login.lower() == organization.lower():
+        course = membership.organization
+        break
+    
+teams = {team.name : team for team in list(course.teams())}
+tas = teams['TAs'].members()
+instructors = teams['Instructors'].members()
+print("Getting team members, please wait...")
+team_members = {team.name: [m.login for m in set(team.members()) - \
+    set(tas) - set(instructors)] for team in teams.values()}
+
+
+### Preprocess
+pairs_df = pd.read_csv(pairs)
+roster_df = pd.read_csv(roster)
+
+header = ['CCID', 'GitHub username', 'expected team', 'current team']
+pairs_df = pairs_df[['Email', 'Assignment 1 Team']]
+pairs_df.columns = ['CCID', 'expected team']
+pairs_df = pairs_df.reindex(columns = header)
+
+roster_df = roster_df[['github_username', 'identifier', 'group_name']]
+roster_df.columns = ['GitHub username', 'CCID', 'expected team']
+
+
+### Check student teams
+pairs_result = pairs_df.copy()
+pairs_result['GitHub username'] = pd.merge(pairs_df, roster_df, on='CCID')['GitHub username_y']
+
+pairs_result['current team'] = pairs_result['GitHub username'].apply(lambda x: \
+    [k for k,v in team_members.items() if x in v]).apply(lambda x: len(x) \
+        and (len(x) == 1 and x[0] or str(x)) or np.nan)
+
+pairs_result = pairs_result.replace(np.nan, '')
+pairs_result = pairs_result.loc[pairs_result['expected team'].str[-3:] != \
+    pairs_result['current team'].str[-3:]].reset_index(drop=True)
+
+print('\n-----\nProblematic students\n-----\n')
+print(pairs_result)
+pairs_result.to_csv('problematic_students.csv', index=False)
+
+
+### Check roster
+roster_result = pd.merge(pairs_df, roster_df, on='CCID', how='outer')
+roster_result = roster_result[['CCID', 'GitHub username_y', 'expected team_x', 'expected team_y']]
+roster_result.columns = header[:3] + ['team in roster']
+roster_result = roster_result.replace(np.nan, '')
+roster_result = roster_result.loc[roster_result['expected team'].str[-3:] != \
+    roster_result['team in roster'].str[-3:]].reset_index(drop=True)
+
+print('\n-----\nRoster\n-----\n')
+print(roster_result)
+roster_result.to_csv('problematic_students_roster.csv', index=False)
```

Comment:
I have a question about the two output files. So from what I can tell, both files are really the same, but one adds the github username too.
There are two ways to find discrepancies:
(1) compare the pairs.csv to the downloaded github classroom roster
(2) compare the pairs.csv to the actual team memberships from the github organization

From the code, I believe #1 corresponds to problematic_students_roster while #2 corresponds to problematic_students? 

Right now, these two seem consistent which is great, because it tells us that the github roster actually reflects the internal team assignments.

That said, Can you maybe add in the readme the difference between these two output files (like one checks against internal github teams while the other against the downloaded roster)?

Thanks again for this! I see that tons of students didn't join any team (I'm tired of reminding students so it's on them at this point) while only one student joined the wrong team while another one joined a team but wasn't officially assigned to any team. So all this has been super useful.

Reply:
you are right, #1 corresponds to problematic_students_roster while #2 corresponds to problematic_students.
I will add this in readme.